### PR TITLE
DM-33414: Allow id to be used in file template

### PR DESCRIPTION
**Pull request by timj**
@gpdf / @TallJimbo this does the job -- I also restrict the id to having to be in the filename itself if it is specified anywhere.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/daf/butler/core/fileTemplates.py [hunk 1]
```diff
@@ -552,9 +558,11 @@ def format(self, ref: DatasetRef) -> str:
...
             )
 
-        # Complain if there's no run
-        if not usedRun:
-            raise KeyError("Template does not include 'run'.")
+        # Complain if there's no run or id
+        if not usedRun and not usedId:
+            missing = ("run" if not usedRun else None, "id" if not usedId else None)
+            text = " or ".join(f"'{m}'" for m in missing if m is not None)
+            raise KeyError(f"Template does not include {text}.")
```
Comment (TallJimbo) [inline python/lsst/daf/butler/core/fileTemplates.py]: Is there a similar check for the dataset type name that could also be relaxed?

Comment (timj) [reply python/lsst/daf/butler/core/fileTemplates.py]: It doesn't look like we require the dataset type to be in the template. "run" was the only thing we required from what I could see (and I was able to use a file template of just `{id}`.

Comment (timj) [reply python/lsst/daf/butler/core/fileTemplates.py]: The fact that this line has no coverage implies that we catch the no-run logic higher up and none of this is needed any more.

Comment (TallJimbo) [reply python/lsst/daf/butler/core/fileTemplates.py]: Sounds like we should just remove it, then, and update the docs to say it's expected?

**Review by TallJimbo — APPROVED** [review]
Seems fine - functionality is clearly what we want, but I'm a little worried that the template validation code is getting a little hard to follow now that we have so many rules (especially if we don't really understand why some of the new lines aren't in the coverage report).

File: python/lsst/daf/butler/core/fileTemplates.py (hunk 1 above)
Comment (timj) [reply python/lsst/daf/butler/core/fileTemplates.py]: I think that this block of code was an early attempt at template validation but at some point we moved the validation logic earlier and never noticed that this block was irrelevant. I'll remove it.
